src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Repository/LocationMetaRepository.cs

```diff
@@ -226,35 +239,74 @@ private static List<GeographicLevel> ListLocationLevels(IReadOnlySet<string> all
         {
             GeographicLevel.LocalAuthority => new LocationLocalAuthorityOptionMeta
             {
-                PublicId = string.Empty,
                 Label = label,
                 Code = row[LocalAuthorityCsvColumns.NewCode],
                 OldCode = row[LocalAuthorityCsvColumns.OldCode]
             },
             GeographicLevel.School => new LocationSchoolOptionMeta
             {
-                PublicId = string.Empty,
                 Label = label,
                 Urn = row[SchoolCsvColumns.Urn],
                 LaEstab = row[SchoolCsvColumns.LaEstab]
             },
             GeographicLevel.Provider => new LocationProviderOptionMeta
             {
-                PublicId = string.Empty,
                 Label = label,
                 Ukprn = row[ProviderCsvColumns.Ukprn]
             },
-            GeographicLevel.RscRegion => new LocationRscRegionOptionMeta
-            {
-                PublicId = string.Empty,
-                Label = label
-            },
+            GeographicLevel.RscRegion => new LocationRscRegionOptionMeta { Label = label },
             _ => new LocationCodedOptionMeta
             {
-                PublicId = string.Empty,
                 Label = label,
                 Code = row[cols.Codes.First()]
             }
         };
     }
+
+    private async Task<PublicIdMappings> CreatePublicIdMappings(
+        DataSetVersion dataSetVersion,
+        CancellationToken cancellationToken)
+    {
+        var mappings = await EntityFrameworkQueryableExtensions
+            .SingleOrDefaultAsync(publicDataDbContext
+                    .DataSetVersionMappings,
+                mapping => mapping.TargetDataSetVersionId == dataSetVersion.Id,
+                cancellationToken);
+
+        if (mappings is null)
+        {
+            return new PublicIdMappings();
+        }
+
+        var mappingsByLevel = mappings
+            .LocationMappingPlan
+            .Levels
+            .ToDictionary(
+                keySelector: level => level.Key,
+                elementSelector: level => level
+                    .Value
+                    .Mappings
+                    .Values
+                    .Where(mapping => mapping.Type is MappingType.AutoMapped or MappingType.ManualMapped)
+                    .ToDictionary(
+                        keySelector: mapping => mapping.CandidateKey!,
+                        elementSelector: mapping => mapping.PublicId));
+
+        return new PublicIdMappings { Levels = mappingsByLevel };
+    }
+
+    private record PublicIdMappings
+    {
+        public Dictionary<GeographicLevel, Dictionary<string, string>> Levels { get; init; } = [];
+
+        public string? GetPublicIdForCandidate(GeographicLevel level, string candidateKey)
+        {
+            if (!Levels.TryGetValue(level, out var mappingLevel))
+            {
+                return null;
+            }
+
+            return mappingLevel.GetValueOrDefault(candidateKey);
```

Comment:
Similar to earlier comment - could be simplified to:

```cs
return Levels.GetValueOrDefault(level)
    ?.GetValueOrDefault(candidateKey);
```

Reply:
Nice - done ta